File: apps/api/src/modules/plan/plan.controller.ts
import { Controller, Get, Post, Put, Delete, Body, Param, ParseUUIDPipe, UseGuards } from '@nestjs/common';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { AuthGuard } from '@nestjs/passport';
import { SuperAdminGuard } from '../../common/guards';
import { PlanService } from './plan.service';
import { CreatePlanDto } from './dto/create-plan.dto';
import { UpdatePlanDto } from './dto/update-plan.dto';

@ApiTags('Super Admin - Plans')
@ApiBearerAuth()
@UseGuards(AuthGuard('jwt'), SuperAdminGuard)
@Controller('super-admin/plans')
export class PlanController {
  constructor(private readonly planService: PlanService) {}

  @Post()
  @ApiOperation({ summary: 'Create a plan' })
  create(@Body() dto: CreatePlanDto) {
    return this.planService.create(dto);
  }

  @Get()
  @ApiOperation({ summary: 'List all plans' })
  findAll() {
    return this.planService.findAll();
  }

  @Get(':id')
  @ApiOperation({ summary: 'Get plan by ID' })
  findById(@Param('id', ParseUUIDPipe) id: string) {
    return this.planService.findById(id);
  }

  @Put(':id')
  @ApiOperation({ summary: 'Update a plan' })
  update(@Param('id', ParseUUIDPipe) id: string, @Body() dto: UpdatePlanDto) {
    return this.planService.update(id, dto);
  }

  @Delete(':id')
  @ApiOperation({ summary: 'Delete a plan' })
  remove(@Param('id', ParseUUIDPipe) id: string) {
    return this.planService.remove(id);
  }

  @Put(':id/modules')
  @ApiOperation({ summary: 'Assign modules to a plan' })
  assignModules(
    @Param('id', ParseUUIDPipe) id: string,
    @Body() body: { module_ids: string[] },
  ) {
    return this.planService.assignModules(id, body.module_ids);
  }
}

File: apps/api-java/src/main/java/br/com/menufacil/controller/PlanController.java
package br.com.menufacil.controller;

import br.com.menufacil.config.security.RequirePermissions;
import br.com.menufacil.dto.AssignPlanModulesRequest;
import br.com.menufacil.dto.CreatePlanRequest;
import br.com.menufacil.dto.PlanResponse;
import br.com.menufacil.service.PlanService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@Tag(name = "Super Admin - Planos", description = "Gerenciamento global de planos de assinatura")
@RestController
@RequestMapping("/super-admin/plans")
@RequiredArgsConstructor
public class PlanController {

    private final PlanService planService;

    @Operation(summary = "Listar todos os planos")
    @RequirePermissions("plan:read")
    @GetMapping
    public ResponseEntity<List<PlanResponse>> findAll() {
        return ResponseEntity.ok(planService.findAll());
    }

    @Operation(summary = "Buscar plano por ID")
    @RequirePermissions("plan:read")
    @GetMapping("/{id}")
    public ResponseEntity<PlanResponse> findById(@PathVariable UUID id) {
        return ResponseEntity.ok(planService.findById(id));
    }

    @Operation(summary = "Criar plano")
    @RequirePermissions("plan:create")
    @PostMapping
    public ResponseEntity<PlanResponse> create(@Valid @RequestBody CreatePlanRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(planService.create(request));
    }

    @Operation(summary = "Atualizar plano")
    @RequirePermissions("plan:update")
    @PutMapping("/{id}")
    public ResponseEntity<PlanResponse> update(
            @PathVariable UUID id,
            @Valid @RequestBody CreatePlanRequest request) {
        return ResponseEntity.ok(planService.update(id, request));
    }

    @Operation(summary = "Remover plano")
    @RequirePermissions("plan:delete")
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable UUID id) {
        planService.delete(id);
        return ResponseEntity.noContent().build();
    }

    @Operation(summary = "Atribuir módulos ao plano (substitui a lista atual)")
    @RequirePermissions("plan:update")
    @PutMapping("/{id}/modules")
    public ResponseEntity<PlanResponse> assignModules(
            @PathVariable UUID id,
            @Valid @RequestBody AssignPlanModulesRequest request) {
        return ResponseEntity.ok(planService.assignModules(id, request.getModuleIds()));
    }
}
